### DISCUSSION TITLE
Setting a cookieless authentication token managed by React context in an authorization header?

### DISCUSSION THREAD

**Discussion by matthewvolk** [trpc/trpc#3687] [Q&A] [answered]
(Noob question alert)

With a focus on privacy, browsers have recently started phasing out support for the 3rd party cookies. This has caused issues for embedded apps for e-commerce platforms like BigCommerce which until now have implemented authentication with the use of 3rd party cookies.

One solution many e-commerce app developers have designed looks like the following:

1. Server authenticates user
2. Server redirects to user to client (for example, Next.js) adding a query parameter to the redirect that looks something like `/?token=<JWT>`
3. Client parses the token from the URL and stores it in React context

This results in client-side token management being held in React context. For example, a component might look like this:

```tsx
export const Component: React.FC = () => {
  const token = useToken(); // React context

  return <pre>JWT: {token}</pre>;
}
```

I'm trying to figure out if it's at all possible to somehow grab the token and access it inside of the config callback [as recommended in the tRPC docs](https://trpc.io/docs/header).

Obviously, I can't use a React hook like `useToken` inside of `createTRPCNext`, and so I'm currently stuck doing something like the following whenever I need to make an authenticated request from a component:

```tsx
export const Component: React.FC = () => {
  const token = useToken(); // React context

  const hello = trpc.sayHelloAuthenticated.useQuery({ token });

  return <h1>{hello.isSuccess && hello.data}</h1>;
}
```

This works, but it's really repetitive to have to add `token` as an input parameter for each authenticated route.

I might be missing a super simple solution here, but I just can't figure out how I could get access to React.context for client-side tRPC requests.

It looks like it might have been possible in [v9 with the HOC approach](https://trpc.io/docs/v9/header), but I'm not sure if this is recommended since I'm already using v10.

If anyone has any suggestions, it would be super appreciated 🙏  I'm not married to the JWT-in-query-parameter approach, and can switch to a completely different method if there's a working solution that might be easier than this.

**Comment by msalsbery** [comment, marked as answer]
You can make your token accessible from anywhere just like any other variables or functions. export/import…it’s just javascript, not everything has to be in the context of React components ;)

**Reply by matthewvolk** [reply]
Thanks Mark! Your reply gave me the idea to parse the JWT query param from the same file where I'm calling `createTRPCNext`. It's been working perfectly 😄
